src/main/kotlin/com/tradeshift/blayze/features/Multinomial.kt

```diff
@@ -37,77 +37,63 @@ class Multinomial private constructor(
+        val outcomesCopy = outcomeToIdx.toMutableMap()
 
-        fun getOrCreateIndex(key: Outcome) = outcomesCopy.getOrPut(key, { outcomesCopy.size })
+        fun getOrCreateIndex(key: Outcome) = outcomesCopy.getOrPut(key) { outcomesCopy.size }
 
         val formattedUpdates = invertUpdates(sampleUpdates(updates.asSequence()))
         for ((feature, counter) in formattedUpdates) {
-            val indexToUpdate = counter.mapKeys { getOrCreateIndex(it.key)  }
+            val indexToUpdate = counter.mapKeys { getOrCreateIndex(it.key) }
             val vec = SparseIntVector.fromMap(indexToUpdate)
             featuresCopy[feature] = featuresCopy[feature]?.add(vec) ?: vec
         }
         return Multinomial(includeFeatureProbability, pseudoCount, outcomesCopy, featuresCopy)
     }
 
-    /**
-     * The unnormalized log probability for each outcome, which is defined as:
-     *
-     *      logProb_o = sum_f(log(S_of + pseudoCount) - log(sum_i(S_oi + pseudoCount))
-     *
-     * where sum_f is a sum over all input features, S_of is the number of times feature f has been seen with outcome o,
-     * and sum_i is a sum over all seen features.
-     *
-     * In the case where the count matrix S_cf is sparse this computation will be dominated by zero values. To retain
-     * performance in this case this implementation reduces the number of computations required by separating
-     * the calculation of zero-valued entries of S and non-zero entries:
-     *
-     *      logProb_o = sum_fnonzero((log(S_of + pseudoCount) - log(sum_i(S_oi + pseudoCount))
-     *                  + sum_fzero((log(0 + pseudoCount) - log(sum_i(S_oi + pseudoCount))
-     *
-     *                = sum_fnonzero((log(S_of + pseudoCount) - log(sum_i(S_oi + pseudoCount))
-     *                  + sum_fzero(1) * (log(S_of + pseudoCount) - log(sum_i(S_oi + pseudoCount))
-     *
-     * The second term only needs to be computed once instead of summing over all input features.
-     */
-    override fun logProbability(outcomes: Set<Outcome>, value: Counter<String>): Map<Outcome, Double> {
-        // sum all log probabilities for non-zero feature-outcome combinations
-        val nonZeroLogProbs = DoubleArray(outcomeIndices.size)
-        val nonZeroCounts = IntArray(outcomeIndices.size)
-        var nFeatures = 0
-        for ((feature, count) in value) {
-            val vector = features[feature]
-            if (vector != null) {
-                nFeatures += count
-                for ((idx, v) in vector) {
-                    nonZeroLogProbs[idx] += count * logProbability(v, nFeaturesPerOutcome[idx])
-                    nonZeroCounts[idx] += count
-                }
-            }
+
+    // See https://en.wikipedia.org/wiki/Dirichlet-multinomial_distribution#Dirichlet-multinomial_as_a_compound_distribution
+    override fun logPosteriorPredictive(outcomes: Set<Outcome>, value: Counter<String>): Map<Outcome, Double> {
```

Comment:
Maybe keep the documentation here that explains why this peculiar implementation leads to performance gains which might not be obvious at first glance.